src/Microsoft.AspNetCore.Server.Kestrel/Internal/Infrastructure/MemoryPool.cs

```diff
@@ -60,20 +61,44 @@ public class MemoryPool : IDisposable
         /// </summary>
         private bool _disposedValue = false; // To detect redundant calls
 
+#if DEBUG
+        public MemoryPoolBlock Lease([CallerMemberName] string memberName = "",
+                             [CallerFilePath] string sourceFilePath = "",
+                             [CallerLineNumber] int sourceLineNumber = 0)
+        {
+            Debug.Assert(!_disposedValue, "Block being leased from disposed pool!");
+#else
+
         /// <summary>
         /// Called to take a block from the pool.
         /// </summary>
         /// <returns>The block that is reserved for the called. It must be passed to Return when it is no longer being used.</returns>
         public MemoryPoolBlock Lease()
         {
+#endif
             MemoryPoolBlock block;
             if (_blocks.TryDequeue(out block))
             {
                 // block successfully taken from the stack - return it
+#if DEBUG
+                block.Leaser = memberName + ", " + sourceFilePath + ", " + sourceLineNumber;
+                block.IsLeased = true;
+#if !NETSTANDARD1_3
+                block.StackTrace = new StackTrace(true).ToString();
+#endif
+#endif
                 return block;
             }
             // no blocks available - grow the pool
-            return AllocateSlab();
+            block = AllocateSlab();
+#if DEBUG
+            block.Leaser = memberName + ", " + sourceFilePath + ", " + sourceLineNumber;
+            block.IsLeased = true;
+#if !NETSTANDARD1_3
+            block.StackTrace = new StackTrace(true).ToString();
```

Comment:
Any difference to Environment.StackTrace? I think that'd be available on both plats.